File: server/src/apis/LobbyAPI.ts
import { Namespace, Socket } from "socket.io";
import { GameID } from "../services/GamesService";
import { LobbyController, LobbyState } from "../services/LobbyService";
import { Server } from "../server";
import { Result } from "../utils";

const EVENT = {
    CONNECT: "connect",
    DISCONNECT: "disconnect",
    TOGGLE_READY: "lobby:ready:toggle",
    ERROR: "lobby:error"
}

export class LobbyAPI {
    private server: Server;
    private id: string;
    private nsp: Namespace;
    private controller: LobbyController;

    constructor(id: string, controller: LobbyController) {
        this.server = Server.getInstance();
        this.id = id;
        this.nsp = this.server.createNameSpace(`lobbies/${this.id}`);
        this.controller = controller;
    }

    listen() {
        this.nsp.on(EVENT.CONNECT, (socket: Socket) => {
            this.controller.addPlayer(socket.id);
            socket.on(EVENT.DISCONNECT, () => {
                this.controller.removePlayer(socket.id);
            })

            socket.on(EVENT.TOGGLE_READY, () => {
                // toggle player's ready status
                this.controller.toggleReady(socket.id);
            })
        })
    }


    startGame(data: GameID) {
        const result = {data}
        this.nsp.emit("lobby:game:start", result);
    }

    updateState(data: LobbyState) {
        const result: Result<LobbyState> = {data};
        this.nsp.emit("lobby:state:update", result)
    }
}

File: app/src/main/java/com/example/liars_dice/api/LobbyAPI.java
package com.example.liars_dice.api;

import java.net.URISyntaxException;

import io.socket.client.IO;
import io.socket.client.Socket;
import io.socket.emitter.Emitter;

public class LobbyAPI {
    private static LobbyAPI INSTANCE = null;

    public enum LobbyEvent {
        UPDATE("lobby:state:update"),
        DISCONNECT("disconnect"),
        READY("lobby:ready:toggle"),
        ALLREADY("lobby:game:start");

        private final String text;

        LobbyEvent(String text) {
            this.text = text;
        }

        @Override
        public String toString() {
            return this.text;
        }
    }

    private Socket socket = null;

    public static LobbyAPI getInstance() {
        if (INSTANCE == null) {
            INSTANCE = new LobbyAPI();
        }
        return INSTANCE;
    }

    public void connect(String id) {
        System.out.println("Connecting with id " + id);
        try {
            socket = IO.socket("http://10.0.2.2:3000/lobbies/"+id);
        } catch (URISyntaxException e) {
            e.printStackTrace();
        }
        socket.connect();
    }

    public void on(LobbyEvent event, Emitter.Listener listener) {
        this.socket.on(event.toString(), listener);
    }

    public void off(LobbyEvent event, Emitter.Listener listener) {
        this.socket.off(event.toString(), listener);
    }

    public void toggleReady() {
        this.socket.emit("lobby:ready:toggle");
    }
}
